### PR TITLE
Fix dataflow of ternary expressions

### PR DESCRIPTION
**Pull request by d367wang**
typetools PR 5000 https://github.com/eisop/checker-framework/pull/151/commits/48f2652bc8bf4801b2e750cd92325583939f2f52 added synthetic variables for ternary expressions to the CFG.
This broke how the Nullness Checker handles ternary expressions, leading to false positives.

This PR fixes handling of ternary expressions by using a special AssignmentNode that does not merge stores, for the assignment of the synthetic variable used in a ternary expression.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationTransfer.java
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/initialization/InitializationTransfer.java:159]: Yesterday we had discussed moving the assert here. Is handling both kinds of store necessary? Is an existing test failing without this or should we add a new test case?

Comment (wmdietl) [reply checker/src/main/java/org/checkerframework/checker/initialization/InitializationTransfer.java:159]: Or should it check for `!n.shouldMergeStore()` like in the super method?

File: dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AssignmentNode.java
Comment (wmdietl) [inline dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AssignmentNode.java:33]:
```suggestion
    /** Whether the then-store and else-store should be merged. */
```

Comment (wmdietl) [reply dataflow/src/main/java/org/checkerframework/dataflow/cfg/node/AssignmentNode.java:33]: Add a comment about this in the class doc and to the dataflow manual.

**Review by wmdietl — CHANGES_REQUESTED** [review]
Thanks for the PR! A few small comments. We will not merge this PR, but some of these changes will go into a new PR once we merged the ternary changes. This is to see whether CI will pass.

File: checker/src/main/java/org/checkerframework/checker/initialization/InitializationTransfer.java
Comment (d367wang) [reply checker/src/main/java/org/checkerframework/checker/initialization/InitializationTransfer.java:159]: Right now only ternary expression may cause conditional transfer result, e.g. `condExpr#num0 = (obj instanceof List) `. In such cases, the LHS is never a `FieldAccess`, so we can only handle regular transfer result here.

**Comment by wmdietl** [comment]
Closed for #165

**Pull request closed without merging**
